### PR TITLE
Fix "missing 'price' property in 'offers'" error in Google Search Console

### PR DESCRIPTION
**Pull request by Aljullu**
### Changes proposed in this Pull Request:

In https://github.com/woocommerce/woocommerce/pull/52105 we updated the structured data of sale prices. That was done by moving the sale price and the regular price under `priceSpecification`, which allowed for both prices to be listed in the metadata. That was an SEO enhancement.

Even though the code from that PR passed the validator checks, after testing it in an online store, it caused Google Search Console to output an error and a warning:

Missing 'price' in 'offers' | Missing 'priceValidUntil' in 'offers'
--- | ---
Captura de pantalla de 2024-11-15 11-02-52 | Captura de pantalla de 2024-11-15 11-02-44

The first one is specially serious, as it might cause the products to be delisted from Google enriched results.

To be honest, I'm not 100% sure why the errors appear, but I think:

* The first one was caused by `priceCurrency` still being under `offers`. I think in some way Google expects that if there is a `priceCurrency` property, there must be a `price` property at the same level. I made it so `priceCurrency` is only added if we detect a `price`, `lowPrice` or `highPrice` properties.
*  The second one seems to indicate that even though we added `validThrough` to the specific `priceSpecification`s, we still needed to keep `priceValidUntil` in the parent `offers`. I added it back.

I also made sure that we don't include `priceType` in the current offer. As this is what's advised by Google. And also made sure that in Grouped products there is no chance that the sale and regular price is the same. (I left comments inline explaining everything).

I tested it in a live store and Google Search Console seems to be fine again.

cc @zhyian in case you also want to take a look. :slightly_smiling_face: 

Note: I'm targeting `release/9.5` as I think this deserves a CFE so we don't ship a regression in WC 9.5.

### How to test the changes in this Pull Request:

1. Create an online store that can be crawled by Google Search Console. Create several products: simple, variable, grouped, on sale, etc.
2. Inspect all product pages using the Search Console URL [Inspector Tool](https://support.google.com/webmasters/answer/9012289?hl=en).
3. Verify there are no errors in any of them.

### CODE REVIEW

**Review by dinhtungdu — COMMENTED** [review]
I can confirm that this PR fix the `missing 'price' property in 'offers'` error.

About the second one, I can't reproduct the warning about missing `priceValidUntil`. I tried `trunk` and your branch (removing the line 373th) and there was no warning about it. Instead, I got two warnings:

image

I used a local store and https://search.google.com/test/rich-results, not sure if it produces different result.

**Comment by Aljullu** [comment]
> I can't reproduct the warning about missing `priceValidUntil`

That's weird, I was able to reproduce consistently. Heads-up that this warning appeared at the product level, not the merchant level like some of the others.

> Instead, I got two warnings:

You're right! We already have two issues for them: https://github.com/woocommerce/woocommerce/issues/41060 and https://github.com/woocommerce/woocommerce/issues/41061.

---

@dinhtungdu would you mind taking another look? I completely rewrote this PR from scratch. After thinking about it twice, I consider that removing `priceCurrency` from the root level might be problematic for product types created by extensions, which might be relying on it. Instead, I kept it there and added back `price` and `priceValidUntil` properties, which were removed in https://github.com/woocommerce/woocommerce/pull/52105. I think this is a safer approach and it seems to be fine for Google Search Console.

File: plugins/woocommerce/includes/class-wc-structured-data.php [hunk 1]
```diff
@@ -234,7 +234,6 @@ public function generate_product_data( $product = null ) {
 						'priceSpecification' => array(
 							array(
 								'@type'                 => 'UnitPriceSpecification',
-								'priceType'             => 'https://schema.org/ListPrice',
```
Comment (Aljullu) [inline plugins/woocommerce/includes/class-wc-structured-data.php:237]: While this wasn't causing an error, according to Google:

> Don't mark the current, offer price with a priceType property.

https://developers.google.com/search/docs/appearance/structured-data/merchant-listing#sale-pricing-example

File: plugins/woocommerce/includes/class-wc-structured-data.php [hunk 2]
```diff
@@ -304,46 +304,55 @@ public function generate_product_data( $product = null ) {
...
-							'price'                 => wc_format_decimal( $min_price, wc_get_price_decimals() ),
-							'priceCurrency'         => $currency,
-							'valueAddedTaxIncluded' => wc_prices_include_tax(),
-							'validThrough'          => $price_valid_until,
-						),
+						$unit_price_specification,
 					),
 				);
 
-				if ( $product->is_on_sale() ) {
```
Comment (Aljullu) [inline plugins/woocommerce/includes/class-wc-structured-data.php:321]: There was the chance that the a Groupped product could have:

1. Product A with regular price 18.
2. Product B with regular price 20 and sale price 18.

That caused the structured data to show the regular price and the sale price as 18. I added a check to make sure the prices are different.

File: plugins/woocommerce/includes/class-wc-structured-data.php [hunk 3]
```diff
@@ -304,46 +303,55 @@ public function generate_product_data( $product = null ) {
...
 					'priceSpecification' => array(
-						array(
-							'@type'                 => 'UnitPriceSpecification',
-							'priceType'             => 'https://schema.org/ListPrice',
-							'price'                 => wc_format_decimal( $min_price, wc_get_price_decimals() ),
-							'priceCurrency'         => $currency,
-							'valueAddedTaxIncluded' => wc_prices_include_tax(),
-							'validThrough'          => $price_valid_until,
-						),
+						$unit_price_specification,
```
Comment (Aljullu) [inline plugins/woocommerce/includes/class-wc-structured-data.php:321]: Same as before, let's make sure we don't include `priceType` in the current offer.

File: plugins/woocommerce/includes/class-wc-structured-data.php [hunk 4]
```diff
@@ -304,46 +303,55 @@ public function generate_product_data( $product = null ) {
...
 					'priceSpecification' => array(
-						array(
-							'@type'                 => 'UnitPriceSpecification',
-							'priceType'             => 'https://schema.org/ListPrice',
-							'price'                 => wc_format_decimal( $product->get_regular_price(), wc_get_price_decimals() ),
-							'priceCurrency'         => $currency,
-							'valueAddedTaxIncluded' => wc_prices_include_tax(),
-							'validThrough'          => $price_valid_until,
-						),
+						$unit_price_specification,
```
Comment (Aljullu) [inline plugins/woocommerce/includes/class-wc-structured-data.php:354]: And here too. :slightly_smiling_face:

File: plugins/woocommerce/includes/class-wc-structured-data.php [hunk 5]
```diff
@@ -304,46 +303,55 @@ public function generate_product_data( $product = null ) {
...
 					),
 				);
 
-				if ( $product->is_on_sale() ) {
+				if ( $product->is_on_sale() && $min_price !== $min_sale_price ) {
```
Comment (Aljullu) [inline plugins/woocommerce/includes/class-wc-structured-data.php:325]: There was the chance that a Grouped product could have:

1. Product A with regular price 18.
2. Product B with regular price 20 and sale price 18.

That caused the structured data to show the regular price and the sale price as 18. I added a check to make sure the prices are different.

File: plugins/woocommerce/includes/class-wc-structured-data.php [hunk 6]
```diff
@@ -370,15 +377,23 @@ public function generate_product_data( $product = null ) {
...
 				),
 			);
+			if (
+				( ! empty( $markup_offer['price'] ) ||
+					! empty( $markup_offer['lowPrice'] ) ||
+					! empty( $markup_offer['highPrice'] )
+				) && empty( $markup_offer['priceCurrency'] )
+			) {
+				$markup_offer['priceCurrency'] = $currency;
+			}
```
Comment (Aljullu) [inline plugins/woocommerce/includes/class-wc-structured-data.php:396]: Only include `priceCurrency` if there is a `price` in the markup offer.

File: plugins/woocommerce/includes/class-wc-structured-data.php [hunk 7]
```diff
@@ -370,15 +377,23 @@ public function generate_product_data( $product = null ) {
 			}
 
 			$markup_offer += array(
-				'priceCurrency' => $currency,
-				'availability'  => 'http://schema.org/' . $stock_status_schema,
-				'url'           => $permalink,
-				'seller'        => array(
+				'priceValidUntil' => $sale_price_valid_until ?? $price_valid_until,
```
Comment (Aljullu) [inline plugins/woocommerce/includes/class-wc-structured-data.php:380]: Add missing `priceValidUntil`.

**Review by dinhtungdu — APPROVED** [review]
Thanks for updating and going the extra mile on this one! The single block page passed the validation in various scenarios in my test. Great work! :shipit:

**Review by dinhtungdu — APPROVED** [review]
Thanks for updating and going the extra mile on this one! The single block page passed the validation in various scenarios in my test. Great work! :shipit:

**Review by mdperez86 — APPROVED** [review]
I tested this PR as well and I was not able to see any error in Google Search Console.

Nice job @Aljullu !!!

**Comment by mklusak** [comment]
Hi, I updated my site to WooCommerce 9.5.1 and this exact problem popped up in my Search Console :-( Next day, dozens of URLs reporting "Missing field 'price'" ... how is that possible?
screen

**Comment by Aljullu** [comment]
@mklusak would you mind sharing the relevant part of the page markup (you can search for `<script type="application/ld+json">`) or a link to the page so we can analyze in detail? Also better if you create a new issue (feel free to ping me in it :slightly_smiling_face: ), as it's easy to lose track of conversations in closed PRs.

And it might be obvious, but I would recommend trying with plugins and custom code disabled to make sure there is nothing else modifying the default logic.

Thanks in advance!

**Comment by mklusak** [comment]
Hello @Aljullu, anonymized ld+json part follows. It's "simple" product.

`{"@context":"https:\/\/schema.org\/","@type":"Product","@id":"https:\/\/www.exmaple.com\/produkt\/xy\/#product","name":"XY","url":"https:\/\/www.exmaple.com\/produkt\/xy\/","description":"....","image":"https:\/\/www.exmaple.com\/wp-content\/uploads\/....jpg","sku":"...","offers":[{"@type":"Offer","priceSpecification":{"0":{"@type":"UnitPriceSpecification","price":"249.00","priceCurrency":"CZK","valueAddedTaxIncluded":false,"validThrough":"2026-12-31"},"priceCurrency":"CZK"},"priceValidUntil":"2026-12-31","availability":"http:\/\/schema.org\/OutOfStock","url":"https:\/\/www.exmaple.com\/produkt\/xy\/","seller":{"@type":"Organization","name":"...","url":"https:\/\/www.exmaple.com"}}]}]}`

I am not sure about "0" index in priceSpecification. Looks suspicious.

**Comment by Aljullu** [comment]
> I am not sure about "0" index in priceSpecification. Looks suspicious.

Yes, this looks weird, the structure should be:

```
      "priceSpecification": [
        {
          "@type": "UnitPriceSpecification",
          "price": "249.00",
          "priceCurrency": "CZK",
          "valueAddedTaxIncluded": false,
          "validThrough": "2026-12-31"
        }
      ],
```

but for some reason you get:

```
      "priceSpecification": {
        "0": {
          "@type": "UnitPriceSpecification",
          "price": "249.00",
          "priceCurrency": "CZK",
          "valueAddedTaxIncluded": false,
          "validThrough": "2026-12-31"
        },
        "priceCurrency": "CZK"
      },
```

Unfortunately, I'm not able to reproduce. :confused: I would suggest doing [conflict testing](https://woocommerce.com/document/woocommerce-self-service-guide/#plugin-and-theme-conflicts) to make sure you don't have any plugins or custom code interfering with it. Unfortunately, if we can't reproduce it on our end, there is not much we can do.

**Comment by mklusak** [comment]
Hi @Aljullu, conflict testing is not a simple task with live site. But I did it and found out that plugin [FOX currency switcher](https://wordpress.org/plugins/woocommerce-currency-switcher/) was to blame. It changes structured data to comply with currency setting etc, but it's not compatible with this recent "offer" structure change. I have reported this error to the plugin authors and am awaiting further action. 

Thank you for your help! I apologize for cluttering this thread with an unrelated topic, but it was the only place I found mention of changes in structured data possibly related to my problem.

**Comment by Aljullu** [comment]
No worries, @mklusak! Thanks for investigating it and reporting the error to the plugin authors! :raised_hands:

**Pull request merged**